File: app/src/main/java/com/example/mvvm_artichecture_sample/ui/MainViewModel.java
package com.example.mvvm_artichecture_sample.ui;

import android.app.Application;

import androidx.lifecycle.MutableLiveData;

import com.example.mvvm_artichecture_sample.R;
import com.example.mvvm_artichecture_sample.base.BaseViewModel;
import com.example.mvvm_artichecture_sample.base.model.MessageModel;
import com.example.mvvm_artichecture_sample.base.network.baseobserver.ErrorConsumer;
import com.example.mvvm_artichecture_sample.base.network.baseobserver.SuccessConsumer;
import com.example.mvvm_artichecture_sample.base.network.baseobserver.WebServiceListener;
import com.example.mvvm_artichecture_sample.base.network.model.APIError;
import com.example.mvvm_artichecture_sample.data.MainRepository;
import com.example.mvvm_artichecture_sample.data.remote.apimodel.CountryModel;
import com.example.mvvm_artichecture_sample.data.remote.apimodel.ReposnseModel;

import java.util.List;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

public class MainViewModel extends BaseViewModel<MainRepository> implements WebServiceListener {
	
	
	private MutableLiveData<Boolean> showLoadig = new MutableLiveData<>();
	private MutableLiveData<MessageModel> showMessage = new MutableLiveData<>();
	private MutableLiveData<List<CountryModel>> countryList = new MutableLiveData<>();
	private static final String MAIN_REQUEST = "main_list";
	private MutableLiveData<String> showToast = new MutableLiveData<>();
	private List<CountryModel> list;
	
	
	public MainViewModel(MainRepository repository, Application application) {
		super(application, repository);
	}
	
	public void fechList() {
		if (isNetworkConnected()) {
			showMessage(false, "", "");
			showLoading(true);
			addDisposable(getRepository().getList().subscribeOn(Schedulers.io())
					              .observeOn(AndroidSchedulers.mainThread())
					              .subscribe(new SuccessConsumer(this, MAIN_REQUEST), new ErrorConsumer(this, MAIN_REQUEST)));
		} else {
			showMessage(true, getString(R.string.error), getString(R.string.internet_error));
		}
		
		
	}
	
	private void setList(List<CountryModel> list) {
		countryList.setValue(list);
	}
	
	public MutableLiveData<List<CountryModel>> updateList() {
		return countryList;
	}
	
	private void showMessage(boolean show, String title, String message) {
		showMessage.setValue(new MessageModel(show, title, message));
	}
	
	public MutableLiveData<MessageModel> message() {
		return showMessage;
	}
	
	private void showLoading(boolean show) {
		showLoadig.setValue(show);
	}
	
	public MutableLiveData<Boolean> loading() {
		return showLoadig;
	}
	
	@Override
	public void onSuccess(Object object, int statusCode, String requestType) throws Exception {
		showLoading(false);
		list = ((ReposnseModel) object).getCountryModel();
		setList(list);
	}
	
	public MutableLiveData<String> showToast() {
		return showToast;
	}
	
	@Override
	public void onError(APIError apiError, int statusCode, String requestType) throws Exception {
		showLoading(false);
		showMessage(true, getString(R.string.error), apiError != null ? apiError.getMessage() : getString(R.string.errorRecieveData));
	}
	
	public void itemClicked(int position) {
		if (list != null)
			showToast.setValue(String.valueOf(list.get(position).getCode()));
	}
}

File: app/src/main/java/com/example/mvvm_artichecture_sample/ui/MainViewModel.kt
package com.example.mvvm_artichecture_sample.ui

import android.app.Application
import androidx.lifecycle.MutableLiveData
import com.example.mvvm_artichecture_sample.R
import com.example.mvvm_artichecture_sample.base.BaseViewModel
import com.example.mvvm_artichecture_sample.base.model.MessageModel
import com.example.mvvm_artichecture_sample.base.network.Output
import com.example.mvvm_artichecture_sample.data.MainRepository
import com.example.mvvm_artichecture_sample.data.remote.apimodel.Country
import com.example.mvvm_artichecture_sample.data.remote.apimodel.ResponsModel
import kotlinx.coroutines.launch

class MainViewModel(repository: MainRepository, application: Application) : BaseViewModel<MainRepository>(application, repository) {

    private val showLoading = MutableLiveData<Boolean>()
    private val showMessage = MutableLiveData<MessageModel>()
    private val countryList = MutableLiveData<List<Country>>()
    private val showToast = MutableLiveData<String>()
    private var list: List<Country>? = null
    private var output: Output<ResponsModel>? = null

    init {
        fetchList()
    }

    private fun fetchList() {
        if (isNetworkConnected) {

            showMessage(false, "", "")

            scope.launch {

                showLoading(true)

                callListApi()

                manageResponse(output!!)

                showLoading(false)
            }

        } else {

            showMessage(true, getString(R.string.error), getString(R.string.internet_error))

        }
    }

    private fun manageResponse(output: Output<ResponsModel>) {


        when (output) {

            is Output.Success -> {
                manageSuccessResponse(output)
            }

            is Output.Error -> {
                manageErrorResponse(output)
            }

        }

    }

    private suspend fun callListApi() {
        output = repository!!.countries(MAIN_REQUEST)
    }

    private fun manageErrorResponse(output: Output.Error) {
        showMessage(true, getString(R.string.error), output.apiError.message!!)
    }

    private fun manageSuccessResponse(output: Output.Success<ResponsModel>) {
        list = output.response.result
        setList(list!!)
    }

    private fun setList(list: List<Country>) {

        countryList.postValue(list)
    }

    fun updateList(): MutableLiveData<List<Country>> {
        return countryList
    }

    private fun showMessage(show: Boolean, title: String, message: String) {
        showMessage.postValue(MessageModel(show, title, message))
    }

    fun message(): MutableLiveData<MessageModel> {
        return showMessage
    }

    private fun showLoading(show: Boolean) {
        showLoading.postValue(show)
    }

    fun loading(): MutableLiveData<Boolean> {
        return showLoading
    }

    fun itemClicked(position: Int) {
        if (list != null) {
            showToast.value = list!![position].code
        }
    }

    fun showToast(): MutableLiveData<String> {
        return showToast
    }

    companion object {
        private const val MAIN_REQUEST = "main_list"
    }
}
